src/hotspot/share/gc/shenandoah/shenandoahAgeCensus.cpp

```diff
@@ -0,0 +1,383 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "gc/shenandoah/mode/shenandoahGenerationalMode.hpp"
+#include "gc/shenandoah/shenandoahAgeCensus.hpp"
+#include "gc/shenandoah/shenandoahHeap.inline.hpp"
+
+ShenandoahAgeCensus::ShenandoahAgeCensus() {
+  assert(ShenandoahHeap::heap()->mode()->is_generational(), "Only in generational mode");
+  if (ShenandoahGenerationalMinTenuringAge > ShenandoahGenerationalMaxTenuringAge) {
+    vm_exit_during_initialization(
+      err_msg("ShenandoahGenerationalMinTenuringAge=" SIZE_FORMAT
+              " should be no more than ShenandoahGenerationalMaxTenuringAge=" SIZE_FORMAT,
+              ShenandoahGenerationalMinTenuringAge, ShenandoahGenerationalMaxTenuringAge));
+  }
+
+  _global_age_table = NEW_C_HEAP_ARRAY(AgeTable*, MAX_SNAPSHOTS, mtGC);
+  CENSUS_NOISE(_global_noise = NEW_C_HEAP_ARRAY(ShenandoahNoiseStats, MAX_SNAPSHOTS, mtGC);)
+  _tenuring_threshold = NEW_C_HEAP_ARRAY(uint, MAX_SNAPSHOTS, mtGC);
+
+  for (int i = 0; i < MAX_SNAPSHOTS; i++) {
+    // Note that we don't now get perfdata from age_table
+    _global_age_table[i] = new AgeTable(false);
+    CENSUS_NOISE(_global_noise[i].clear();)
+    // Sentinel value
+    _tenuring_threshold[i] = MAX_COHORTS;
+  }
+  if (ShenandoahGenerationalAdaptiveTenuring && !ShenandoahGenerationalCensusAtEvac) {
+    size_t max_workers = ShenandoahHeap::heap()->max_workers();
+    _local_age_table = NEW_C_HEAP_ARRAY(AgeTable*, max_workers, mtGC);
+    CENSUS_NOISE(_local_noise = NEW_C_HEAP_ARRAY(ShenandoahNoiseStats, max_workers, mtGC);)
+    for (uint i = 0; i < max_workers; i++) {
+      _local_age_table[i] = new AgeTable(false);
+      CENSUS_NOISE(_local_noise[i].clear();)
+    }
+  } else {
+    _local_age_table = nullptr;
+  }
+  _epoch = MAX_SNAPSHOTS - 1;  // see update_epoch()
+}
+
+CENSUS_NOISE(void ShenandoahAgeCensus::add(uint obj_age, uint region_age, uint region_youth, size_t size, uint worker_id) {)
+NO_CENSUS_NOISE(void ShenandoahAgeCensus::add(uint obj_age, uint region_age, size_t size, uint worker_id) {)
+  if (obj_age <= markWord::max_age) {
+    assert(obj_age < MAX_COHORTS && region_age < MAX_COHORTS, "Should have been tenured");
+#ifdef SHENANDOAH_CENSUS_NOISE
```

Comment:
The idea was to have this be independently controllable:
```
#ifndef PRODUCT
// Enable noise instrumentation
#define SHENANDOAH_CENSUS_NOISE 1
#endif  // PRODUCT

#ifdef SHENANDOAH_CENSUS_NOISE

#define CENSUS_NOISE(x) x
#define NO_CENSUS_NOISE(x)
...
#else  // SHENANDOAH_CENSUS_NOISE
#define CENSUS_NOISE(x)
#define NO_CENSUS_NOISE(x) x
#endif // SHENANDOAH_CENSUS_NOISE
```

The instrumentation is present only in non-product (non-release) builds. It's thus tested ony in debug/fastdebug mode, but can be turned on, if needed in release mode when doing specific debugging, by means of editing the above.

Since both paths are tested (one in fastdebug mode and the other in release mode), all code paths are always tested and won't bit-rot.

Reply:
ok, then